File: SICP/Lecture_2/Code/Java/CompoundData/src/fayelab/sicp/compounddata/vec/Vec.java
package fayelab.sicp.compounddata.vec;

import static java.util.Arrays.asList;

import java.util.List;

public class Vec
{
    public static List<Integer> makeVec(int xcor, int ycor)
    {
        return asList(xcor, ycor);
    }

    public static int xcor(List<Integer> vec)
    {
        return head(vec);
    }

    public static int ycor(List<Integer> vec)
    {
        return tail(vec);
    }

    public static List<Integer> add(List<Integer> vec1, List<Integer> vec2)
    {
        return makeVec(xcor(vec1) + xcor(vec2), ycor(vec1) + ycor(vec2));
    }

    public static List<Integer> scale(int c, List<Integer> vec)
    {
        return makeVec(c * xcor(vec), c * ycor(vec));
    }
    
    private static int head(List<Integer> pair)
    {
        return pair.get(0);
    }
    
    private static int tail(List<Integer> pair)
    {
        return pair.get(1);
    }
}

File: SICP/Lecture_2/Code/Python/compound_data/vector/vec.py
def make_vec(_xcor, _ycor):
    return [_xcor, _ycor]

def xcor(vec):
    return head(vec)

def ycor(vec):
    return tail(vec)

def add(vec1, vec2):
    return make_vec(xcor(vec1) + xcor(vec2), ycor(vec1) + ycor(vec2))

def scale(c, vec):
    return make_vec(c * xcor(vec), c * ycor(vec))

def head(pair):
    return pair[0]

def tail(pair):
    return pair[1]